File: OximyMac/App/AppState.swift
import Foundation
import SwiftUI

/// Global application state
@MainActor
final class AppState: ObservableObject {

    // MARK: - App Phase
    // Flow: enrollment → setup → ready

    enum Phase: String {
        case enrollment // Step 1: Connect to workspace (6-digit code)
        case setup      // Step 2: Install cert + enable proxy
        case ready      // Done: Monitoring active
    }

    @Published var phase: Phase = .enrollment {
        didSet {
            guard oldValue != phase else { return }
            SentryService.shared.addStateBreadcrumb(
                category: "app.phase",
                message: "Phase changed",
                data: ["from": oldValue.rawValue, "to": phase.rawValue]
            )
            SentryService.shared.updateContext(
                phase: phase.rawValue,
                proxyEnabled: isProxyEnabled,
                port: currentPort
            )
        }
    }

    // MARK: - Connection Status

    enum ConnectionStatus {
        case disconnected
        case connecting
        case connected
        case error(String)

        var color: Color {
            switch self {
            case .disconnected: return .gray
            case .connecting: return .orange
            case .connected: return .green
            case .error: return .red
            }
        }

        var label: String {
            switch self {
            case .disconnected: return "Disconnected"
            case .connecting: return "Connecting..."
            case .connected: return "Monitoring Active"
            case .error(let msg): return msg
            }
        }

        var isConnected: Bool {
            if case .connected = self { return true }
            return false
        }
    }

    @Published var connectionStatus: ConnectionStatus = .disconnected

    // MARK: - Permissions Status

    @Published var isCertificateInstalled: Bool = false
    @Published var isProxyEnabled: Bool = false

    // MARK: - Stats

    @Published var eventsCapturedToday: Int = 0
    @Published var currentPort: Int = Constants.preferredPort

    // MARK: - Account

    @Published var workspaceName: String = ""
    @Published var deviceId: String = ""
    @Published var isLoggedIn: Bool = false

    // MARK: - Main Tabs

    enum MainTab: String, CaseIterable {
        case home = "Home"
        case settings = "Settings"

        var icon: String {
            switch self {
            case .home: return "house.fill"
            case .settings: return "gearshape.fill"
            }
        }
    }

    @Published var selectedTab: MainTab = .home

    // MARK: - Device Info

    var deviceName: String {
        Host.current().localizedName ?? "Mac"
    }

    // MARK: - Initialization

    init() {
        loadPersistedState()
        setupNotificationObservers()
    }

    private func setupNotificationObservers() {
        NotificationCenter.default.addObserver(
            forName: .workspaceNameUpdated,
            object: nil,
            queue: .main
        ) { [weak self] notification in
            guard let self = self else { return }
            if let newName = notification.object as? String {
                print("[AppState] NOTIFICATION: workspaceNameUpdated to '\(newName)'")
                // Use Task to update on MainActor (compatible with macOS 13+)
                Task { @MainActor in
                    self.workspaceName = newName
                }
            }
        }
    }

    // MARK: - State Management

    private func loadPersistedState() {
        let defaults = UserDefaults.standard

        // DEBUG: Log all relevant UserDefaults at startup
        let storedWorkspace = defaults.string(forKey: Constants.Defaults.workspaceName)
        let storedToken = defaults.string(forKey: Constants.Defaults.deviceToken)
        let storedDeviceId = defaults.string(forKey: Constants.Defaults.deviceId)
        let storedWorkspaceId = defaults.string(forKey: Constants.Defaults.workspaceId)
        print("[AppState] INIT - UserDefaults state:")
        print("[AppState]   workspaceName: '\(storedWorkspace ?? "nil")'")
        print("[AppState]   deviceToken: \(storedToken != nil ? "***" : "nil")")
        print("[AppState]   deviceId: '\(storedDeviceId ?? "nil")'")
        print("[AppState]   workspaceId: '\(storedWorkspaceId ?? "nil")'")

        // Check for MDM-managed configuration FIRST
        let mdmConfig = MDMConfigService.shared
        print("[AppState] isManagedDevice: \(mdmConfig.isManagedDevice)")

        if mdmConfig.isManagedDevice {
            print("[AppState] MDM-managed device detected")

            // Apply MDM configuration to standard UserDefaults
            mdmConfig.applyManagedConfiguration()

            // Handle managed device token
            if let token = mdmConfig.managedDeviceToken {
                writeDeviceTokenFile(token)
            }

            // Set account info from MDM if available
            if let workspaceName = mdmConfig.managedWorkspaceName {
                print("[AppState] Setting workspace from MDM: '\(workspaceName)'")
                self.workspaceName = workspaceName
                isLoggedIn = true
            }

            if let deviceId = mdmConfig.managedDeviceId {
                self.deviceId = deviceId
            }

            // Determine phase for managed device
            if mdmConfig.shouldSkipAllSetup {
                // MDM says skip everything - go directly to ready
                print("[AppState] MDM: Skipping all setup, going to ready phase")
                phase = .ready
                OximyLogger.shared.isSetupComplete = true
                SentryService.shared.updateSetupStatus(complete: true)
                startServices()
                return
            } else if mdmConfig.shouldSkipEnrollment {
                // MDM says skip enrollment only - go to setup
                print("[AppState] MDM: Skipping enrollment, going to setup phase")
                phase = .setup
                return
            }
            // Fall through to normal logic if MDM doesn't specify phase
        }

        // Standard (non-MDM) state loading
        // Load account info if available
        if let workspace = defaults.string(forKey: Constants.Defaults.workspaceName) {
            print("[AppState] Setting workspace from UserDefaults: '\(workspace)'")
            workspaceName = workspace
            isLoggedIn = true
        }
        if let device = defaults.string(forKey: Constants.Defaults.deviceId) {
            deviceId = device
        }

        // Determine phase based on what's been completed
        let hasDeviceToken = defaults.string(forKey: Constants.Defaults.deviceToken) != nil
        let setupComplete = defaults.bool(forKey: Constants.Defaults.setupComplete)

        // Ensure device token file exists (handles upgrade from older versions)
        if hasDeviceToken {
            ensureDeviceTokenFile()
        }

        if hasDeviceToken && setupComplete {
            // Both enrollment and setup done
            phase = .ready
            OximyLogger.shared.isSetupComplete = true
            SentryService.shared.updateSetupStatus(complete: true)
            startServices()
        } else if hasDeviceToken {
            // Enrolled but setup not complete
            phase = .setup
        } else {
            // Fresh install - start with enrollment
            phase = .enrollment
        }
    }

    // MARK: - Phase Transitions

    /// Called after successful enrollment (step 1 complete)
    func completeEnrollment() {
        OximyLogger.shared.log(.ENROLL_STATE_101, "Enrollment complete", data: [
            "workspace": workspaceName,
            "device_id": deviceId
        ])
        // Move to setup phase
        phase = .setup
    }

    /// Called when both cert and proxy are enabled (step 2 complete)
    func completeSetup() {
        UserDefaults.standard.set(true, forKey: Constants.Defaults.setupComplete)
        phase = .ready
        connectionStatus = .connected
        OximyLogger.shared.isSetupComplete = true
        SentryService.shared.updateSetupStatus(complete: true)
        OximyLogger.shared.log(.APP_STATE_101, "Setup complete, entering ready phase")
        startServices()
    }

    /// Go back to enrollment from setup
    func goBackToEnrollment() {
        phase = .enrollment
    }

    /// Skip setup and go to ready state without enabling proxy
    func skipSetup() {
        UserDefaults.standard.set(true, forKey: Constants.Defaults.setupComplete)
        phase = .ready
        connectionStatus = .disconnected
        startServices()
    }

    /// Handle auth failure after retries exhausted
    func handleAuthFailure() {
        logout()
    }

    /// Start background services
    private func startServices() {
        NSLog("[AppState] startServices() called")
        HeartbeatService.shared.start()
        SyncService.shared.start()

        // Start the proxy when entering ready phase
        // Use a slight delay to ensure app initialization is complete
        NSLog("[AppState] Scheduling MITMService start...")
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
            Task { @MainActor in
                NSLog("[AppState] Starting MITMService now...")
                do {
                    try await MITMService.shared.start()
                    NSLog("[AppState] MITMService started successfully")
                } catch {
                    NSLog("[AppState] Failed to start MITMService: %@", String(describing: error))
                    OximyLogger.shared.log(.MITM_FAIL_304, "MITM process start failed in startServices", data: [
                        "error": error.localizedDescription
                    ], err: (type: "MITMError", code: "MITM_START_FAILED", message: error.localizedDescription))
                }
            }
        }
    }

    /// Stop background services
    private func stopServices() {
        HeartbeatService.shared.stop()
        SyncService.shared.stop()
    }

    /// Check if setup requirements are met
    var isSetupComplete: Bool {
        isCertificateInstalled && isProxyEnabled
    }

    /// Reset everything (for debugging/logout)
    func reset() {
        stopServices()

        let defaults = UserDefaults.standard
        defaults.removeObject(forKey: Constants.Defaults.setupComplete)
        defaults.removeObject(forKey: Constants.Defaults.workspaceName)
        defaults.removeObject(forKey: Constants.Defaults.deviceToken)
        defaults.removeObject(forKey: Constants.Defaults.deviceId)
        defaults.removeObject(forKey: Constants.Defaults.workspaceId)

        // Delete device token file
        deleteDeviceTokenFile()

        // Reset setup status for re-enrollment tracking
        OximyLogger.shared.isSetupComplete = false
        SentryService.shared.updateSetupStatus(complete: false)

        phase = .enrollment
        connectionStatus = .disconnected
        workspaceName = ""
        deviceId = ""
        isLoggedIn = false
        isCertificateInstalled = false
        isProxyEnabled = false
    }

    // MARK: - Device Token File Management

    /// Write device token to ~/.oximy/device-token for the Python addon
    @discardableResult
    private func writeDeviceTokenFile(_ token: String) -> Bool {
        let tokenPath = Constants.deviceTokenPath
        let oximyDir = Constants.oximyDir
        let fm = FileManager.default

        do {
            // Ensure ~/.oximy directory exists
            if !fm.fileExists(atPath: oximyDir.path) {
                try fm.createDirectory(at: oximyDir, withIntermediateDirectories: true)
            }

            // Write token as plain text (addon uses .strip() on read)
            try token.write(to: tokenPath, atomically: true, encoding: .utf8)

            // Set file permissions to owner-only readable (0600) for security
            try fm.setAttributes([.posixPermissions: 0o600], ofItemAtPath: tokenPath.path)

            print("[AppState] Device token written to \(tokenPath.path)")
            return true
        } catch {
            print("[AppState] Failed to write device token file: \(error.localizedDescription)")
            SentryService.shared.captureError(error, context: [
                "operation": "write_device_token",
                "path": tokenPath.path
            ])
            return false
        }
    }

    /// Delete the device token file on logout/reset
    @discardableResult
    private func deleteDeviceTokenFile() -> Bool {
        let tokenPath = Constants.deviceTokenPath
        let fm = FileManager.default

        guard fm.fileExists(atPath: tokenPath.path) else {
            return true  // File doesn't exist, nothing to delete
        }

        do {
            try fm.removeItem(at: tokenPath)
            print("[AppState] Device token file deleted")
            return true
        } catch {
            print("[AppState] Failed to delete device token file: \(error.localizedDescription)")
            SentryService.shared.captureError(error, context: [
                "operation": "delete_device_token",
                "path": tokenPath.path
            ])
            return false
        }
    }

    /// Ensure device token file exists if we have a token (handles app upgrades)
    private func ensureDeviceTokenFile() {
        guard let token = UserDefaults.standard.string(forKey: Constants.Defaults.deviceToken),
              !token.isEmpty else {
            return
        }

        let tokenPath = Constants.deviceTokenPath
        let fm = FileManager.default

        // Check if file already exists with correct content
        if fm.fileExists(atPath: tokenPath.path),
           let existingToken = try? String(contentsOf: tokenPath, encoding: .utf8),
           existingToken.trimmingCharacters(in: .whitespacesAndNewlines) == token {
            return  // File exists with correct content
        }

        // File missing or has stale content - write it
        writeDeviceTokenFile(token)
    }

    // MARK: - Account

    func login(workspaceName: String, deviceToken: String, deviceId: String?, workspaceId: String?) {
        let defaults = UserDefaults.standard
        defaults.set(workspaceName, forKey: Constants.Defaults.workspaceName)
        defaults.set(deviceToken, forKey: Constants.Defaults.deviceToken)

        // Write token to file for Python addon
        writeDeviceTokenFile(deviceToken)

        if let deviceId = deviceId {
            defaults.set(deviceId, forKey: Constants.Defaults.deviceId)
            self.deviceId = deviceId
        }

        if let workspaceId = workspaceId {
            defaults.set(workspaceId, forKey: Constants.Defaults.workspaceId)
        }

        self.workspaceName = workspaceName
        self.isLoggedIn = true

        // Update Sentry user context with full identity
        SentryService.shared.setFullUserContext(
            workspaceName: workspaceName,
            deviceId: deviceId,
            workspaceId: workspaceId
        )

        OximyLogger.shared.log(.AUTH_AUTH_001, "User logged in", data: [
            "workspace": workspaceName,
            "has_device_id": deviceId != nil
        ])
    }

    /// Check if logout is allowed (not blocked by MDM)
    var canLogout: Bool {
        !MDMConfigService.shared.disableUserLogout
    }

    func logout() {
        // Check if logout is blocked by MDM
        if MDMConfigService.shared.disableUserLogout {
            print("[AppState] Logout blocked by MDM policy")
            return
        }

        // Stop services first
        stopServices()

        let defaults = UserDefaults.standard
        defaults.removeObject(forKey: Constants.Defaults.workspaceName)
        defaults.removeObject(forKey: Constants.Defaults.deviceToken)
        defaults.removeObject(forKey: Constants.Defaults.deviceId)
        defaults.removeObject(forKey: Constants.Defaults.workspaceId)
        defaults.removeObject(forKey: Constants.Defaults.setupComplete)

        // Delete device token file
        deleteDeviceTokenFile()

        OximyLogger.shared.log(.AUTH_AUTH_002, "User logged out")

        // Clear Sentry user context (after log so the event retains user attribution)
        SentryService.shared.clearUser()

        // Reset setup status AFTER logging logout (so the event isn't suppressed)
        OximyLogger.shared.isSetupComplete = false
        SentryService.shared.updateSetupStatus(complete: false)

        workspaceName = ""
        deviceId = ""
        isLoggedIn = false

        // Return to enrollment (step 1)
        phase = .enrollment
    }
}

File: OximyWindows/src/OximyWindows/Core/AppState.cs
using System.ComponentModel;
using System.IO;
using System.Runtime.CompilerServices;

namespace OximyWindows.Core;

/// <summary>
/// Global application state management.
/// Implements singleton pattern with observable properties.
/// Matches the Mac app's three-phase flow: Enrollment -> Setup -> Ready
/// </summary>
public class AppState : INotifyPropertyChanged
{
    private static AppState? _instance;
    public static AppState Instance => _instance ??= new AppState();

    private Phase _phase = Phase.Enrollment;
    public Phase Phase
    {
        get => _phase;
        private set => SetProperty(ref _phase, value);
    }

    private ConnectionStatus _connectionStatus = ConnectionStatus.Disconnected;
    public ConnectionStatus ConnectionStatus
    {
        get => _connectionStatus;
        set => SetProperty(ref _connectionStatus, value);
    }

    private string _deviceName = Environment.MachineName;
    public string DeviceName
    {
        get => _deviceName;
        set => SetProperty(ref _deviceName, value);
    }

    private string _workspaceName = string.Empty;
    public string WorkspaceName
    {
        get => _workspaceName;
        set => SetProperty(ref _workspaceName, value);
    }

    private string _workspaceId = string.Empty;
    public string WorkspaceId
    {
        get => _workspaceId;
        private set => SetProperty(ref _workspaceId, value);
    }

    private string _deviceId = string.Empty;
    public string DeviceId
    {
        get => _deviceId;
        private set => SetProperty(ref _deviceId, value);
    }

    private string _deviceToken = string.Empty;
    public string DeviceToken
    {
        get => _deviceToken;
        private set => SetProperty(ref _deviceToken, value);
    }

    private int _currentPort = Constants.PreferredPort;
    public int CurrentPort
    {
        get => _currentPort;
        set => SetProperty(ref _currentPort, value);
    }

    private int _eventsCapturedToday;
    public int EventsCapturedToday
    {
        get => _eventsCapturedToday;
        set => SetProperty(ref _eventsCapturedToday, value);
    }

    private int _eventsPending;
    public int EventsPending
    {
        get => _eventsPending;
        set => SetProperty(ref _eventsPending, value);
    }

    private DateTime? _lastSyncTime;
    public DateTime? LastSyncTime
    {
        get => _lastSyncTime;
        set => SetProperty(ref _lastSyncTime, value);
    }

    private string _errorMessage = string.Empty;
    public string ErrorMessage
    {
        get => _errorMessage;
        set => SetProperty(ref _errorMessage, value);
    }

    private bool _isSetupCertificateComplete;
    public bool IsSetupCertificateComplete
    {
        get => _isSetupCertificateComplete;
        set => SetProperty(ref _isSetupCertificateComplete, value);
    }

    private bool _isSetupProxyComplete;
    public bool IsSetupProxyComplete
    {
        get => _isSetupProxyComplete;
        set => SetProperty(ref _isSetupProxyComplete, value);
    }

    private AppState()
    {
        LoadPersistedState();
    }

    /// <summary>
    /// Load persisted state from settings.
    /// </summary>
    private void LoadPersistedState()
    {
        var settings = Properties.Settings.Default;

        DeviceName = string.IsNullOrEmpty(settings.DeviceName)
            ? Environment.MachineName
            : settings.DeviceName;

        // Load saved credentials
        if (!string.IsNullOrEmpty(settings.DeviceToken))
        {
            DeviceToken = settings.DeviceToken;
            DeviceId = settings.DeviceId;
            WorkspaceName = settings.WorkspaceName;
            WorkspaceId = settings.WorkspaceId;

            // Always go to Connected if we have a token (setup happens in background)
            Phase = Phase.Connected;

            // Ensure token file exists for Python addon (handles upgrade from older versions)
            WriteDeviceTokenFile(settings.DeviceToken);
        }
        else
        {
            Phase = Phase.Enrollment;
        }

        // Update events count
        EventsCapturedToday = Constants.CountTodayEvents();
    }

    /// <summary>
    /// Complete enrollment with device registration response.
    /// Transitions directly to Connected phase (setup happens in background).
    /// </summary>
    public void CompleteEnrollment(string deviceId, string deviceToken, string workspaceName, string workspaceId)
    {
        var settings = Properties.Settings.Default;
        settings.DeviceId = deviceId;
        settings.DeviceToken = deviceToken;
        settings.WorkspaceName = workspaceName;
        settings.WorkspaceId = workspaceId;
        settings.SetupComplete = true;  // Mark setup complete immediately
        settings.Save();

        DeviceId = deviceId;
        DeviceToken = deviceToken;
        WorkspaceName = workspaceName;
        WorkspaceId = workspaceId;
        Phase = Phase.Connected;  // Go directly to Connected (skip Setup)

        // Write device token to file for Python addon to read
        WriteDeviceTokenFile(deviceToken);
    }

    /// <summary>
    /// Write device token to ~/.oximy/device-token file for Python addon.
    /// </summary>
    private static void WriteDeviceTokenFile(string token)
    {
        try
        {
            var tokenPath = Path.Combine(Constants.OximyDir, "device-token");
            Directory.CreateDirectory(Constants.OximyDir);
            File.WriteAllText(tokenPath, token);
            System.Diagnostics.Debug.WriteLine($"[AppState] Device token written to {tokenPath}");
        }
        catch (Exception ex)
        {
            System.Diagnostics.Debug.WriteLine($"[AppState] Failed to write device token file: {ex.Message}");
        }
    }

    /// <summary>
    /// Mark certificate setup as complete.
    /// </summary>
    public void CompleteCertificateSetup()
    {
        IsSetupCertificateComplete = true;
        CheckSetupComplete();
    }

    /// <summary>
    /// Mark proxy setup as complete.
    /// </summary>
    public void CompleteProxySetup()
    {
        IsSetupProxyComplete = true;
        CheckSetupComplete();
    }

    /// <summary>
    /// Check if all setup steps are complete and transition to Connected phase.
    /// </summary>
    private void CheckSetupComplete()
    {
        if (IsSetupCertificateComplete && IsSetupProxyComplete)
        {
            var settings = Properties.Settings.Default;
            settings.SetupComplete = true;
            settings.Save();

            Phase = Phase.Connected;
        }
    }

    /// <summary>
    /// Complete the entire setup and transition to Connected phase.
    /// </summary>
    public void CompleteSetup()
    {
        var settings = Properties.Settings.Default;
        settings.SetupComplete = true;
        settings.Save();

        IsSetupCertificateComplete = true;
        IsSetupProxyComplete = true;
        Phase = Phase.Connected;
    }

    /// <summary>
    /// Skip setup for now (set up later).
    /// </summary>
    public void SkipSetup()
    {
        Phase = Phase.Connected;
    }

    /// <summary>
    /// Complete login with workspace credentials (legacy method for compatibility).
    /// </summary>
    public void CompleteLogin(string workspaceName, string deviceToken)
    {
        var settings = Properties.Settings.Default;
        settings.WorkspaceName = workspaceName;
        settings.DeviceToken = deviceToken;
        settings.Save();

        WorkspaceName = workspaceName;
        DeviceToken = deviceToken;
        Phase = Phase.Connected;
    }

    /// <summary>
    /// Log out and return to enrollment phase.
    /// </summary>
    public void Logout()
    {
        var settings = Properties.Settings.Default;
        settings.DeviceToken = string.Empty;
        settings.DeviceId = string.Empty;
        settings.WorkspaceName = string.Empty;
        settings.WorkspaceId = string.Empty;
        settings.SetupComplete = false;
        settings.Save();

        DeviceToken = string.Empty;
        DeviceId = string.Empty;
        WorkspaceName = string.Empty;
        WorkspaceId = string.Empty;
        IsSetupCertificateComplete = false;
        IsSetupProxyComplete = false;
        Phase = Phase.Enrollment;

        // Remove device token file
        DeleteDeviceTokenFile();
    }

    /// <summary>
    /// Delete the device token file on logout.
    /// </summary>
    private static void DeleteDeviceTokenFile()
    {
        try
        {
            var tokenPath = Path.Combine(Constants.OximyDir, "device-token");
            if (File.Exists(tokenPath))
            {
                File.Delete(tokenPath);
                System.Diagnostics.Debug.WriteLine($"[AppState] Device token file deleted");
            }
        }
        catch (Exception ex)
        {
            System.Diagnostics.Debug.WriteLine($"[AppState] Failed to delete device token file: {ex.Message}");
        }
    }

    /// <summary>
    /// Reset to initial enrollment state.
    /// </summary>
    public void Reset()
    {
        Logout();
    }

    /// <summary>
    /// Complete the onboarding phase (legacy).
    /// </summary>
    public void CompleteOnboarding()
    {
        Phase = Phase.Enrollment;
    }

    /// <summary>
    /// Complete the permissions phase (legacy).
    /// </summary>
    public void CompletePermissions()
    {
        Phase = Phase.Connected;
    }

    /// <summary>
    /// Reset to initial onboarding state (legacy).
    /// </summary>
    public void ResetOnboarding()
    {
        Reset();
    }

    /// <summary>
    /// Update device name.
    /// </summary>
    public void UpdateDeviceName(string name)
    {
        var settings = Properties.Settings.Default;
        settings.DeviceName = name;
        settings.Save();

        DeviceName = name;
    }

    /// <summary>
    /// Refresh events count.
    /// </summary>
    public void RefreshEventsCount()
    {
        EventsCapturedToday = Constants.CountTodayEvents();
    }

    /// <summary>
    /// Get relative time string for last sync.
    /// </summary>
    public string GetLastSyncRelativeTime()
    {
        if (!LastSyncTime.HasValue)
            return "Never";

        var elapsed = DateTime.UtcNow - LastSyncTime.Value;

        if (elapsed.TotalSeconds < 60)
            return "Just now";
        if (elapsed.TotalMinutes < 60)
            return $"{(int)elapsed.TotalMinutes} min ago";
        if (elapsed.TotalHours < 24)
            return $"{(int)elapsed.TotalHours} hours ago";

        return $"{(int)elapsed.TotalDays} days ago";
    }

    #region INotifyPropertyChanged

    public event PropertyChangedEventHandler? PropertyChanged;

    protected bool SetProperty<T>(ref T field, T value, [CallerMemberName] string? propertyName = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value))
            return false;

        field = value;
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        return true;
    }

    protected void OnPropertyChanged([CallerMemberName] string? propertyName = null)
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
    }

    #endregion
}
